accepted/0000-basic-orchestration.md

```diff
@@ -0,0 +1,125 @@
+
+In the initial version, the scope of the orchestration support will be limited to clustering. In later iterations it can be expanded to support other uses cases.
+
+# Detailed design
+[design]: #detailed-design
+
+The `salt-netapi` library already support calling Salt runner. 
+
+The Uyuni actions must be enhanced to make Salt runner calls besides minions calls.
+
+## Action enhancements.
+
+A runner action won't need a collection of `ServerAction` objects because the action is not executed on a set of minions directly.
+
+Salt supports asynchronous runner calls but it doesn't allow attaching metadata like `action-id`. Therefore the `jid` returned by Salt must be stored in the database in order to match the return event to the originating action. 
+
+### Database and Hibernate changes
+
+In order to store the `jid` and the executions outcome a new table is needed:
+
+```sql
+CREATE TABLE IF NOT EXISTS rhnOrchestrationAction (
+
+    action_id       NUMERIC NOT NULL
+                            CONSTRAINT rhn_actsaltrunjob_aid_fk
+                            REFERENCES rhnAction (id)
+                            ON DELETE CASCADE,
+    jid              VARCHAR(100) NOT NULL,
+    status           NUMERIC NOT NULL
+                         CONSTRAINT rhn_server_action_status_fk
+                             REFERENCES rhnActionStatus (id),
+    result_code      NUMERIC,
+    result_msg       TEXT,
+    pickup_time      TIMESTAMPTZ,
+    completion_time  TIMESTAMPTZ,
+    created         TIMESTAMPTZ
+                        DEFAULT (current_timestamp) NOT NULL,
+    primary key (action_id, jid)
+);
+
+```
+
+A new collection will be added to the cluster actions (`ClusterJoinNodeAction`, `ClusterRemoveNodeAction`, etc):
+
+```java
+class ClusterJoinNodeAction {
+  [...]
+  private Set<ActionSaltRunnerJob> runnerJobs = new HashSet<>();
+  [...]
+}
+```
+
+The Hibernate mappings will need to be adapted:
+
+```xml
+<subclass name="com.redhat.rhn.domain.action.cluster.ClusterJoinNodeAction"
+        lazy="true" discriminator-value="516">
+    [...]
+    <set name="runnerJobs" outer-join="false" cascade="all" lazy="true"
+          inverse="true">
+        <key column="action_id"/>
+        <one-to-many
+                class="com.redhat.rhn.domain.action.ActionSaltRunnerJob" />
+    </set>
+    <join table="rhnActionClusterJoinNode">
+      [...]
+    </join>
+</subclass>
+
+```
+
+### Action result display
+
+Actions executed on a system are shown in the Events tab of the system details. A similar approach should be used for orchestration results. They should be displayed in the details of the target entity or in a generic orchestration UI.
+
+In the case of clusters, the UI should be enhanced to show a view similar to the system Events tab.
```

Comment:
Currently, an event result is nothing more than a status and basically a string with the result itself. Thinking about orchestration, I can imagine we might want a more detailed view of some sort - ie. a single long string might be a bit basic, especially if lots of nodes are involved.

How do you envision such a UI?